File: serializer/src/main/scala/org/beangle/data/serializer/io/path/Path.java
package org.beangle.data.serializer.io.path;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class Path {

    private final String[] chunks;
    private transient String pathAsString;
    private transient String pathExplicit;
    private static final Path DOT = new Path(new String[] {"."});

    public Path(String pathAsString) {
        // String.split() too slow. StringTokenizer too crappy.
        List result = new ArrayList();
        int currentIndex = 0;
        int nextSeparator;
        this.pathAsString = pathAsString;
        while ((nextSeparator = pathAsString.indexOf('/', currentIndex)) != -1) {
            // normalize explicit paths
            result.add(normalize(pathAsString, currentIndex, nextSeparator));
            currentIndex = nextSeparator + 1;
        }
        result.add(normalize(pathAsString,currentIndex,pathAsString.length()));
        String[] arr = new String[result.size()];
        result.toArray(arr);
        chunks = arr;
    }
    
    private String normalize(String s, int start, int end) {
        if (end - start > 3 
                && s.charAt(end-3) == '['
                && s.charAt(end-2) == '1'
                && s.charAt(end-1) == ']') {
            this.pathAsString = null;
            return s.substring(start, end-3);
        } else {
            return s.substring(start, end);
        }
        
    }

    public Path(String[] chunks) {
        this.chunks = chunks;
    }

    public String toString() {
        if (pathAsString == null) {
            StringBuffer buffer = new StringBuffer();
            for (int i = 0; i < chunks.length; i++) {
                if (i > 0) buffer.append('/');
                buffer.append(chunks[i]);
            }
            pathAsString = buffer.toString();
        }
        return pathAsString;
    }

    public String explicit() {
        if (pathExplicit == null) {
            StringBuffer buffer = new StringBuffer();
            for (int i = 0; i < chunks.length; i++) {
                if (i > 0) buffer.append('/');
                String chunk = chunks[i];
                buffer.append(chunk);
                int length = chunk.length();
                if (length > 0) {
                    char c = chunk.charAt(length-1);
                    if (c != ']' && c != '.') {
                        buffer.append("[1]");
                    }
                }
            }
            pathExplicit = buffer.toString();
        }
        return pathExplicit;
    }

    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Path)) return false;

        final Path other = (Path) o;
        if (chunks.length != other.chunks.length) return false;
        for (int i = 0; i < chunks.length; i++) {
            if (!chunks[i].equals(other.chunks[i])) return false;
        }

        return true;
    }

    public int hashCode() {
        int result = 543645643;
        for (int i = 0; i < chunks.length; i++) {
            result = 29 * result + chunks[i].hashCode();
        }
        return result;
    }

    public Path relativeTo(Path that) {
        int depthOfPathDivergence = depthOfPathDivergence(chunks, that.chunks);
        String[] result = new String[chunks.length + that.chunks.length - 2 * depthOfPathDivergence];
        int count = 0;

        for (int i = depthOfPathDivergence; i < chunks.length; i++) {
            result[count++] = "..";
        }
        for (int j = depthOfPathDivergence; j < that.chunks.length; j++) {
            result[count++] = that.chunks[j];
        }

        if (count == 0) {
            return DOT;
        } else {
            return new Path(result);
        }
    }

    private int depthOfPathDivergence(String[] path1, String[] path2) {
        int minLength = Math.min(path1.length, path2.length);
        for (int i = 0; i < minLength; i++) {
            if (!path1[i].equals(path2[i])) {
                return i;
            }
        }
        return minLength;
    }

    public Path apply(Path relativePath) {
    	//FIXME FastStack
        Stack absoluteStack = new Stack();

        for (int i = 0; i < chunks.length; i++) {
            absoluteStack.push(chunks[i]);
        }

        for (int i = 0; i < relativePath.chunks.length; i++) {
            String relativeChunk = relativePath.chunks[i];
            if (relativeChunk.equals("..")) {
                absoluteStack.pop();
            } else if (!relativeChunk.equals(".")) {
                absoluteStack.push(relativeChunk);
            }
        }

        String[] result = new String[absoluteStack.size()];
        for (int i = 0; i < result.length; i++) {
            result[i] = (String) absoluteStack.get(i);
        }

        return new Path(result);
    }
    
    public boolean isAncestor(Path child) {
        if (child == null || child.chunks.length < chunks.length) {
            return false;
        }
        for (int i = 0; i < chunks.length; i++) {
            if (!chunks[i].equals(child.chunks[i])) {
                return false;
            }
        }
        return true;
    }
}

File: serializer/src/main/scala/org/beangle/data/serializer/io/Path.scala
package org.beangle.data.serializer.io

import org.beangle.commons.lang.Strings
object Path {
  val Dot = new Path(Array("."))
}
class Path(val chunks: Array[String]) {

  override def toString: String = {
    Strings.join(chunks, "/")
  }

  def explicit(): String = {
    val buffer = new StringBuffer()
    (0 until chunks.length) foreach { i =>
      if (i > 0) buffer.append('/')
      val chunk = chunks(i)
      buffer.append(chunk)
      val length = chunk.length()
      if (length > 0) {
        val c = chunk.charAt(length - 1)
        if (c != ']' && c != '.') {
          buffer.append("[1]")
        }
      }
    }
    buffer.toString()
  }

  override def equals(o: Any): Boolean = {
    if (this eq o.asInstanceOf[AnyRef]) return true
    o match {
      case other: Path =>
        if (chunks.length != other.chunks.length) return false
        var i = 0
        while (i < chunks.length) {
          if (!chunks(i).equals(other.chunks(i))) return false
          i += 1
        }
        true
      case _ => false
    }
  }

  override def hashCode(): Int = {
    var result = 543645643
    (0 until chunks.length) foreach { i =>
      result = 29 * result + chunks(i).hashCode()
    }
    result
  }

  def relativeTo(that: Path): Path = {
    val depthOfPathDivergence = getDepthOfPathDivergence(chunks, that.chunks)
    val result = new Array[String](chunks.length + that.chunks.length - 2 * depthOfPathDivergence)
    var count = 0

    var i = 0
    (0 until chunks.length) foreach { i =>
      count += 1
      result(count) = ".."
    }
    (depthOfPathDivergence until chunks.length) foreach { i =>
      count += 1
      result(count) = that.chunks(i)
    }

    if (count == 0) Path.Dot else new Path(result)
  }

  private def getDepthOfPathDivergence(path1: Array[String], path2: Array[String]): Int = {
    val minLength = Math.min(path1.length, path2.length)
    var i = 0
    while (i < minLength) {
      if (!path1(i).equals(path2(i))) return i
      i = i + 1
    }
    minLength
  }
  def isAncestor(child: Path): Boolean = {
    if (child == null || child.chunks.length < chunks.length) return false
    var i = 0
    while (i < chunks.length) {
      if (!(chunks(i) == child.chunks(i))) return false
      i = i + 1
    }
    true
  }
}